File: backend/arrlink/api/backup.py
"""DB backup endpoints: list existing backups, trigger one manually."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel

from ..core import backup as backup_core
from ..deps import get_db
from ..state import State
from .auth import CurrentUser

router = APIRouter(prefix="/api/backup", tags=["backup"])


@router.get("")
def list_backups(_user: CurrentUser, request: Request) -> list[dict]:
    settings = request.app.state.settings
    return backup_core.list_backups(settings.backup_dir)


@router.get("/settings")
def get_backup_settings(_user: CurrentUser, request: Request, db: State = Depends(get_db)) -> dict:
    enabled, retention_days, interval_hours = backup_core.effective_backup_settings(
        db, request.app.state.settings
    )
    return {"enabled": enabled, "retention_days": retention_days, "interval_hours": interval_hours}


class BackupSettingsIn(BaseModel):
    enabled: bool
    retention_days: int
    interval_hours: int


@router.put("/settings")
def put_backup_settings(
    body: BackupSettingsIn,
    _user: CurrentUser,
    db: State = Depends(get_db),
) -> dict:
    db.set_setting("backup_enabled", body.enabled)
    db.set_setting("backup_retention_days", body.retention_days)
    db.set_setting("backup_interval_hours", body.interval_hours)
    db.log_event(
        "info",
        f"backup settings updated (enabled={body.enabled}, "
        f"retention_days={body.retention_days}, interval_hours={body.interval_hours})",
    )
    return {
        "enabled": body.enabled,
        "retention_days": body.retention_days,
        "interval_hours": body.interval_hours,
    }


@router.post("/run")
async def run_backup(_user: CurrentUser, request: Request, db: State = Depends(get_db)) -> dict:
    settings = request.app.state.settings
    _enabled, retention_days, _interval_hours = backup_core.effective_backup_settings(db, settings)
    # Manual trigger always runs regardless of `enabled` -- that flag only
    # gates the automatic nightly loop; a manual "run now" is an explicit
    # admin action and should always be honored.
    return await asyncio.to_thread(
        backup_core.run_backup_cycle,
        db,
        settings.db_path,
        settings.backup_dir,
        retention_days,
    )

File: backend-ts/src/api/backup.ts
import type { FastifyInstance } from 'fastify'
import { z } from 'zod'
import type { Settings } from '../config/env.js'
import type { DbClient } from '../db/client.js'
import type { SettingsStore } from '../db/settings-store.js'
import {
  effectiveBackupSettings,
  listBackups,
  runBackupCycle,
  type BackupInfo,
} from '../core/backup.js'
import { logEvent } from '../db/events.js'
import { HttpError } from '../http-error.js'
import { getCurrentUser } from './auth.js'

/** DB backup endpoints. Ported from api/backup.py. Wire format is snake_case,
 * matching web/ and the Python backend. */

export interface BackupRouteOptions {
  db: DbClient
  settingsStore: SettingsStore
  env: Settings
}

const BackupSettingsSchema = z.object({
  enabled: z.boolean(),
  retention_days: z.number().int(),
  interval_hours: z.number().int(),
})

function backupOut(b: BackupInfo): Record<string, unknown> {
  return { name: b.name, size: b.size, created_at: b.createdAt }
}

export function registerBackupRoutes(
  app: FastifyInstance,
  opts: BackupRouteOptions,
): void {
  const { db, settingsStore, env } = opts

  app.get('/api/backup', (request) => {
    getCurrentUser(request, db, settingsStore, env)
    return listBackups(env.backupDir).map(backupOut)
  })

  app.get('/api/backup/settings', (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const { enabled, retentionDays, intervalHours } = effectiveBackupSettings(
      settingsStore,
      env,
    )
    return { enabled, retention_days: retentionDays, interval_hours: intervalHours }
  })

  app.put('/api/backup/settings', (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const parsed = BackupSettingsSchema.safeParse(request.body)
    if (!parsed.success) throw new HttpError(422, 'invalid request body')
    const body = parsed.data
    settingsStore.setSetting('backup_enabled', body.enabled)
    settingsStore.setSetting('backup_retention_days', body.retention_days)
    settingsStore.setSetting('backup_interval_hours', body.interval_hours)
    logEvent(
      db,
      'info',
      `backup settings updated (enabled=${body.enabled}, retention_days=${body.retention_days}, ` +
        `interval_hours=${body.interval_hours})`,
    )
    return body
  })

  app.post('/api/backup/run', async (request) => {
    getCurrentUser(request, db, settingsStore, env)
    // Ignores `enabled` -- that flag only gates the automatic nightly loop.
    const { retentionDays } = effectiveBackupSettings(settingsStore, env)
    return runBackupCycle(db, env.dbPath, env.backupDir, retentionDays)
  })
}
